README.md
# time_difference

A new Flutter project.

## Feature of this package

This project is to find the time elapsed from certain date and represent like Just now, 2 minutes ago and so on similar as facebook status. It takes parameter String format of DateTime.

It returns the date in format as Dec 3, 2020 if the difference is more than 2 days.

The String here is just conversionof DateTime format of DateTime.now() as DateTime.now().toString().

No necessary to convert other formats of DateTime to String because it won't work for other ones.




CHANGELOG.md
### 0.1.3+1
- Updates for Today into Few hours ago

### 0.1.3
- Revise proper example

### 0.1.2
- Changes Function name 

### 0.1.1
- Sets Example properly

### 0.1.0
- Example added.
- Separated package

### 0.0.1
- Adds homepage
- Creates License file
- Returns date difference as ago


test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

//import 'package:time_difference/time_difference.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    
    //await tester.pumpWidget(MyApp());

    // Verify that our counter starts at 0.
    expect(find.text('0'), findsOneWidget);
    expect(find.text('1'), findsNothing);

    // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });
}


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

//import 'package:time__check/main.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    //await tester.pumpWidget(MyApp());

    // Verify that our counter starts at 0.
    expect(find.text('0'), findsOneWidget);
    expect(find.text('1'), findsNothing);

    // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });
}


lib/main.dart
import 'package:flutter/material.dart';
import 'package:time_difference/time_difference.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Text(timeAgo("2020-12-08 19:29:54.47452")),
          ],
        ),
      ),
    );
  }
}


lib/time_difference.dart
import 'package:intl/intl.dart';

String timeAgo(String datetime) {
  DateTime d1 = DateTime.now();
  DateTime d2 = DateTime.parse(datetime);

  //Difference between now and given time in minutes
  var minutes = int.parse((d1.difference(d2)).inMinutes.toString());

  switch (minutes) {
    case 0:
      return "Just Now";
      break;

    case 1:
      return "1 min ago";
      break;

    case 2:
      return "2 mins ago";
      break;

    //If it is more than two minutes return no. of hours.
    default:
      return hourFormatter(datetime);
      break;
  }
}

String hourCalculate(String datetime) {
  DateTime d1 = DateTime.now();
  DateTime d2 = DateTime.parse(datetime);

  var diff = d1.difference(d2); //d2-d1
  return ((diff.inHours).toString());
}

String hourFormatter(String datetime) {
  //Converted time difference in hour
  var hours = int.parse(hourCalculate(datetime));

  switch (hours) {
    case 0:
      return "Few minutes earlier";
      break;

    case 1:
      return "1 hour ago";
      break;

    case 2:
      return "2 hours ago";
      break;

    //Converted into days for more than 2 hours
    default:
      return dayFormatter(datetime);
      break;
  }
}

String dayCalculate(String datetime) {
  DateTime d1 = DateTime.now();
  DateTime d2 = DateTime.parse(datetime);

  var diff = d1.difference(d2); //d2-d1
  return ((diff.inDays).toString());
}

String dayFormatter(String datetime) {
  var days = int.parse(dayCalculate(datetime));

  switch (days) {
    case 0:
      return "Today";
      break;

    case 1:
      return "Yesterday";
      break;

    case 2:
      return "2 days ago";
      break;

    // Return the given time if the difference is more than 2 days.
    default:
      return getDateOnly(datetime);
      break;
  }
}

String getDateOnly(String date) {
  DateTime d1 = DateTime.parse(date);
  return DateFormat.yMMMd().format(d1);
}


example/lib/example.dart
import 'package:flutter/material.dart';
import 'package:time_difference/time_difference.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(title: 'Flutter Time Demo'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
          child: Column(children: [
        Text(DateTime.now().toString()),
        SizedBox(width: 12.0),

        //This time format is the default format (2020-12-08 13:29:54.47452) of DateTime.now()
        //Store in same format and later Compare on same format
        Text(timeAgo("2020-12-08 13:29:54.47452"))
      ])),
    );
  }
}
